trace/kafka/component.go

```diff
@@ -0,0 +1,158 @@
+	WaitForAll RequiredAcks = -1
+)
+
+const fieldSetMsg = "Setting property '%v' for '%v'"
+
+// Builder gathers all required and optional properties, in order
+// to construct a Kafka AsyncProducer.
+type Builder struct {
+	brokers     []string
+	cfg         *sarama.Config
+	chErr       chan error
+	tag         opentracing.Tag
+	enc         encoding.EncodeFunc
+	contentType string
+	errors      []error
+}
+
+// NewBuilder initiates the AsyncProducer builder chain.
+// The builder instantiates the component using default values for
+// EncodeFunc and Content-Type header.
+func NewBuilder(brokers []string) *Builder {
+	cfg := sarama.NewConfig()
+	cfg.Version = sarama.V0_11_0_0
+
+	return &Builder{
+		brokers:     brokers,
+		cfg:         cfg,
+		chErr:       make(chan error),
+		tag:         opentracing.Tag{Key: "type", Value: "async"},
+		enc:         json.Encode,
+		contentType: json.Type,
+		errors:      []error{},
+	}
+}
+
+// WithTimeout sets the dial timeout for the AsyncProducer.
+func (ab *Builder) WithTimeout(dial time.Duration) *Builder {
+	if dial <= 0*time.Second {
+		ab.errors = append(ab.errors, errors.New("dial timeout has to be positive"))
+	} else {
+		ab.cfg.Net.DialTimeout = dial
+		log.Info(fieldSetMsg, "dial timeout", dial)
+	}
+
+	return ab
+}
+
+// WithVersion sets the kafka versionfor the AsyncProducer.
+func (ab *Builder) WithVersion(version string) *Builder {
+	if version == "" {
+		ab.errors = append(ab.errors, errors.New("version is required"))
+		return ab
+	}
+	v, err := sarama.ParseKafkaVersion(version)
+	if err != nil {
+		ab.errors = append(ab.errors, errors.New("failed to parse kafka version"))
+		return ab
+	}
+	log.Info(fieldSetMsg, "version", version)
+	ab.cfg.Version = v
+
+	return ab
+}
+
+// WithRequiredAcksPolicy adjusts how many replica acknowledgements
+// broker must see before responding.
+func (ab *Builder) WithRequiredAcksPolicy(ack RequiredAcks) *Builder {
+	if !isValidRequiredAcks(ack) {
+		ab.errors = append(ab.errors, errors.New("invalid value for required acks policy provided"))
+	} else {
+		log.Info(fieldSetMsg, "required acks", ack)
+		ab.cfg.Producer.RequiredAcks = sarama.RequiredAcks(ack)
+	}
+
+	return ab
+}
+
+// WithEncoder sets a specific encoder implementation and Content-Type string header;
+// if no option is provided it defaults to json.
+func (ab *Builder) WithEncoder(enc encoding.EncodeFunc, contentType string) *Builder {
+	if enc == nil {
+		ab.errors = append(ab.errors, errors.New("encoder is nil"))
+	} else {
+		log.Info(fieldSetMsg, "encoder", enc)
+		ab.enc = enc
+	}
+	if contentType == "" {
+		ab.errors = append(ab.errors, errors.New("content type is empty"))
+	} else {
+		log.Info(fieldSetMsg, "content type", contentType)
+		ab.contentType = contentType
+	}
+
+	return ab
+}
+
+// Create constructs the AsyncProducer component by applying the gathered properties.
+func (ab *Builder) Create() (*AsyncProducer, error) {
+	if len(ab.brokers) == 0 {
```

Comment:
should we maybe check for the brokers in the NewBuilder method ? 
e.g. When the argument is provided ?

Reply:
I had moved this check in `Create()` to keep `NewBuilder` minimal and avoid an extra declaration there; let me know how it looks now 😄 

Also, I'm not sure that this check is required at all, since it's [handled](https://github.com/Shopify/sarama/blob/a20d26723487482b4f1af8b8537657a5ba72ae37/client.go#L138) by Sarama's client one level down, so we're just aborting earlier.